src/Adapter/Product/FeatureValue/Update/ProductFeatureValueUpdater.php

```diff
@@ -0,0 +1,203 @@
+     *
+     * @return FeatureValueId[]
+     *
+     * @throws CannotAddFeatureValueException
+     * @throws CannotUpdateFeatureValueException
+     * @throws CoreException
+     * @throws DBALException
+     * @throws FeatureValueNotFoundException
+     * @throws InvalidArgumentException
+     * @throws FeatureNotFoundException
+     */
+    public function setFeatureValues(ProductId $productId, array $productFeatureValues): array
+    {
+        // First assert that all entities exist
+        $this->productRepository->assertProductExists($productId);
+        foreach ($productFeatureValues as $productFeatureValue) {
+            $this->featureRepository->assertExists($productFeatureValue->getFeatureId());
+            if (null !== $productFeatureValue->getFeatureValueId()) {
+                $this->featureValueRepository->assertExists($productFeatureValue->getFeatureValueId());
+            }
+        }
+
+        foreach ($productFeatureValues as $productFeatureValue) {
+            if (null !== $productFeatureValue->getFeatureValueId()) {
+                $this->updateFeatureValue($productFeatureValue);
+            } else {
+                $this->addFeatureValue($productFeatureValue);
+            }
+        }
+
+        return $this->updateAssociations($productId, $productFeatureValues);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param array $productFeatureValues
+     *
+     * @return FeatureValueId[]
+     *
+     * @throws DBALException
+     * @throws InvalidArgumentException
+     */
+    private function updateAssociations(ProductId $productId, array $productFeatureValues): array
+    {
+        // First delete all associations from the product
+        $this->connection->delete(
+            $this->dbPrefix . 'feature_product',
+            ['id_product' => $productId->getValue()]
+        );
+
+        // Then create all new ones
+        $insertedValues = [];
+        $productFeatureValueIds = [];
+        foreach ($productFeatureValues as $productFeatureValue) {
+            $insertedValues[] = [
+                'id_product' => $productId->getValue(),
+                'id_feature' => $productFeatureValue->getFeatureId()->getValue(),
+                'id_feature_value' => $productFeatureValue->getFeatureValueId()->getValue(),
+            ];
+            $productFeatureValueIds[] = $productFeatureValue->getFeatureValueId();
+        }
+
+        $this->connection->insert($this->dbPrefix . 'feature_product', $insertedValues);
+
+        return $productFeatureValueIds;
+    }
+
+    /**
+     * @param ProductFeatureValue $productFeatureValue
+     *
+     * @throws CannotUpdateFeatureValueException
+     * @throws CoreException
+     * @throws FeatureValueNotFoundException
+     */
+    private function updateFeatureValue(ProductFeatureValue $productFeatureValue): void
+    {
+        // Only custom values need to be updated
+        if (null === $productFeatureValue->getLocalizedCustomValues()) {
```

Comment:
Syntax sugar suggestion 🍭 
```suggestion
        if ($productFeatureValue->isCustom()) {
```

Reply:
I removed the `isCustom` in the VO so I prefer to rely on `getLocalizedCustomValues()`